### DISCUSSION TITLE
Custom Task and access to GeneratePathProperty create property from PackageReference element

### DISCUSSION THREAD

**Discussion by ebfortin** [dotnet/msbuild#8269] [Q&A] [answered]
I'm writing a custom task for MSBuild to help automate configuration of Source Generators. I can get the PackageReference list and there are metadata on it. However it doesn't really give me any hint on the location of the assembly for the package. For example for the FullPath it gives me "C:\PATH_TO_PRJECT\PACKAGE_NAME" which doesn't point to anything.

So I use the GeneratePathProperty="true" on the PackageReference element, expecting a property with nomenclature "PkgSome_Assembly_Name" having the value of the path to the package to be created. However, there's no such property present when I retrieve all the evaluated properties out of the Project.

Anyone can help on this?

**Comment by KalleOlaviNiemitalo** [comment, marked as answer]
Does your custom task get items as parameters from a target, or does it use the MSBuild API to load and evaluate a second instance of the project?  I'd normally expect the former, but the way you "retrieve all the evaluated properties out of the Project" makes me suspect the latter.

Are you trying to locate analyzer assemblies (including source generators) or reference assemblies?

In Microsoft.CSharp.Core.targets, the CoreCompile target passes `References="@(ReferencePathWithRefAssemblies)"` and `Analyzers="@(Analyzer)"` to the Csc task, so these include the DLLs from NuGet packages.

<https://github.com/dotnet/roslyn/blob/15b43b33901c88f68ef43f8314b5a2457716780d/src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets#L59-L138>

<https://source.dot.net/> shows that the FindReferenceAssembliesForReferences target in Microsoft.Common.CurrentVersion.targets creates ReferencePathWithRefAssemblies items from ReferencePath items.

<https://github.com/dotnet/msbuild/blob/2badb37d109910fbd3155cf8743224b7a27494d8/src/Tasks/Microsoft.Common.CurrentVersion.targets#L2390-L2396>

The ResolveReferences target returns the ReferencePath items, so if you make your custom target depend on that, I think the ReferencePath items should have been created by then, and the target could pass them to your custom task.

<https://github.com/dotnet/msbuild/blob/2badb37d109910fbd3155cf8743224b7a27494d8/src/Tasks/Microsoft.Common.CurrentVersion.targets#L1470-L1473>

**Reply by ebfortin** [reply]
Thanks for your answer. It gives me some possible solution.

To answer your question I'm looking at references. I want to automate the cumbersome manual management of dependencies for source generators. Something that looks like this :

```
<ItemGroup>
   <TargetPathWithTargetPlatformMoniker Include="$(PkgHandlebars_Net)\lib\netstandard2.0\Handlebars.dll" IncludeRuntimeDependency="false" />
   <TargetPathWithTargetPlatformMoniker Include="$(PkgMicrosoft_CSharp)\lib\netstandard2.0\Microsoft.CSharp.dll" IncludeRuntimeDependency="false" />
   <TargetPathWithTargetPlatformMoniker Include="$(PkgAntlr4_Runtime_Standard)\lib\netstandard2.0\Antlr4.Runtime.Standard.dll" IncludeRuntimeDependency="false" />

   <None Include="$(OutputPath)\$(AssemblyName).dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />

   <None Include="$(PkgHandlebars_Net)\lib\netstandard2.0\*.dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
   <None Include="$(PkgMicrosoft_CSharp)\lib\netstandard2.0\*.dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" /> 
   <None Include="$(PkgAntlr4_Runtime_Standard)\lib\netstandard2.0\*.dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
</ItemGroup>
```

So the idea is to add a "GeneratorIncluded='true'" to any reference that we want to include in the source generator package. I then iterate through all the reference, gets the one with that property at true and generate the equivalent of the above. I think the info you gave me put me in the right direction. But if you have more thoughts I'm buying.

**Reply by KalleOlaviNiemitalo** [reply]
Oh, so you are packaging Roslyn analyzers, rather than configuring source generators that your projects consume.

I do something similar for packing the outputs of two C# projects to the same NuGet package. The ProjectReference item has Pack="true" metadata. I add the name of a custom target to the TargetsForTfmSpecificBuildOutput property. The custom target copies ReferenceCopyLocalPaths items to BuildOutputInPackage items if they have %(Pack) == 'true'.

AFAIK, there may be conflicts if two Roslyn analyzers depend on different versions of the same assemblies; see <https://github.com/dotnet/roslyn/issues/41785#issuecomment-588434954>. MSBuild tackled similar conflicts in <https://github.com/dotnet/msbuild/issues/1754> by using AssemblyLoadContext, but I suspect it would be difficult for Roslyn to know which assemblies in the [Analyzers parameter](https://github.com/dotnet/roslyn/blob/15b43b33901c88f68ef43f8314b5a2457716780d/src/Compilers/Core/MSBuildTask/ManagedCompiler.cs#L71-L75) belong to the same AssemblyLoadContext.

**Reply by ebfortin** [reply]
Yes I'm packaging Roslyn analyzers.

What you told me earlier made me advance a bit. The only thing I'm missing now is how to return (Output) an array of string from a Custom Task. After that I think I'm good.

**Reply by ebfortin** [reply]
I was finally able to make it work. Still learning on how MSBuild works, so part of it was the learning curve.

Thanks a lot for your help. I would have taken a lot more time if not for you.
